src/libopensc/pkcs15-nqApplet.c

```diff
@@ -0,0 +1,238 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "common/compat_strlcpy.h"
+#include "log.h"
+#include "pkcs15.h"
+#include "cards.h"
+
+static const char name_Card[] = "NQ-Applet";
+static const char name_Vendor[] = "JCOP4";
+
+static int get_nqapplet_certificate(sc_card_t *card, u8 data_id, struct sc_pkcs15_der *cert_info)
+{
+	int rv;
+	u8 buffer[3072];
+	size_t cb_buffer = sizeof(buffer);
+	LOG_FUNC_CALLED(card->ctx);
+
+	rv = sc_get_data(card, data_id, buffer, cb_buffer);
+	LOG_TEST_RET(card->ctx, rv, "GET DATA failed");
+	if (rv == 0)
+	{
+		LOG_TEST_RET(card->ctx, SC_ERROR_FILE_NOT_FOUND, "No certificate data returned");
+	}
+
+	if (cert_info != NULL)
+	{
+		if (cert_info->value != NULL)
+		{
+			free(cert_info->value);
+		}
+		cert_info->value = malloc(rv);
+		if (cert_info->value != NULL)
+		{
+			cert_info->len = rv;
+			memcpy(cert_info->value, buffer, rv);
+		}
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_pin(sc_pkcs15_card_t *p15card, const char *id, u8 reference)
+{
+	int rv;
+	struct sc_pkcs15_auth_info pin_info;
+	struct sc_pkcs15_object pin_obj;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	memset(&pin_info, 0, sizeof(pin_info));
+	memset(&pin_obj, 0, sizeof(pin_obj));
+
+	sc_pkcs15_format_id(id, &pin_info.auth_id);
+	pin_info.auth_type = SC_PKCS15_PIN_AUTH_TYPE_PIN;
+	pin_info.attrs.pin.reference = reference;
+	pin_info.attrs.pin.flags = SC_PKCS15_PIN_FLAG_INITIALIZED | SC_PKCS15_PIN_FLAG_CASE_SENSITIVE | SC_PKCS15_PIN_TYPE_FLAGS_PIN_GLOBAL | SC_PKCS15_PIN_AUTH_TYPE_PIN;
+	pin_info.attrs.pin.type = SC_PKCS15_PIN_TYPE_UTF8;
+	pin_info.attrs.pin.min_length = 6;
+	pin_info.attrs.pin.stored_length = 6;
+	pin_info.attrs.pin.max_length = 6;
+	pin_info.attrs.pin.pad_char = '\0';
+	pin_info.tries_left = -1; //TODO
+	pin_info.max_tries = 3;
+
+	strlcpy(pin_obj.label, "UserPIN", sizeof(pin_obj.label));
+	pin_obj.flags = SC_PKCS15_CO_FLAG_MODIFIABLE | SC_PKCS15_CO_FLAG_PRIVATE;
+
+	rv = sc_pkcs15emu_add_pin_obj(p15card, &pin_obj, &pin_info);
+	LOG_TEST_RET(card->ctx, rv, "sc_pkcs15emu_add_pin_obj failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int add_nqapplet_certificate(sc_pkcs15_card_t *p15card, const char *id, const char *name, u8 data_id)
+{
+	int rv;
+	struct sc_pkcs15_cert_info cert_info;
+	struct sc_pkcs15_object cert_obj;
+	sc_card_t *card = p15card->card;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	memset(&cert_info, 0, sizeof(cert_info));
+	memset(&cert_obj, 0, sizeof(cert_obj));
+
+	sc_pkcs15_format_id(id, &cert_info.id);
+	cert_info.authority = 0;
+	rv = get_nqapplet_certificate(card, data_id, &cert_info.value);
+	LOG_TEST_RET(card->ctx, rv, "Failed to get certificate");
+
+	strlcpy(cert_obj.label, name, sizeof(cert_obj.label));
+	cert_obj.flags = 0;
```

Comment:
As far as I know some compilers initialize a struct defined on the stack with unspecified values and this struct has lots of members. I just wanted to avoid unexpected side effects in sc_pkcs15emu_add_x509_cert.

Reply:
This was not resolved yet. Setting flats to `0` is not needed when you previously called memset on the whole structure. If you would not, it would indeed be needed (ideally for all the members). This way, it will probably get optimized out, but it unneeded step. The same for the `cert_info.authority` above.

Reply:
I suppose it would be optimized out, but anyway, removed the unnecessary code.